### ISSUE TITLE
Finish generator implementation and documentation

### ISSUE THREAD

**Issue by westlicht** [westlicht/performer#12]

**Comment by UncleRoel** [comment]
I don't know the generator code, but while discussing with a friend last week he mentioned a nice idea: it wound be great to have the option to mirror (up/down) and reverse(left/right) sequences, or for example double the note distances, scramble all notes, etc.

**Comment by westlicht** [comment]
I have a few ideas for additional generators. But I'm also still looking for better accessibility of this feature. I'm not super happy with the current implementation but still struggle to come up with something neater.

**Comment by UncleRoel** [comment]
Perhaps -depending on how many there will be- you could add pages to the context menu, with F5 advancing to the next page. This way it's a fast interface that doesn't go to a different menu style.

**Comment by UncleRoel** [comment]
Inserting and deleting steps in a sequence would also be a neat feature to have in the list! Great for varying shifting patterns.

**Comment by madeinspace** [comment]
An lfo generator via a curve track would be pretty handy:  |/\|/\|/\|/\

**Comment by james-everitt** [comment]
Feature add for the Euclidean Gen:
Would be nice to be able to define the OFF and ON values for the selected page with F4 and F5. That would make it useful for things like octave switch or on the retrig page being able to set 'ON' to 2 retrigs/one octave up rather than always 100% of the parameter value

Also random generator based on Perlin 2D noise would be a cool generator. 8x8 that gets unwrapped to 64 steps

**Comment by toneburst** [comment]
> Also random generator based on Perlin 2D noise would be a cool generator. 8x8 that gets unwrapped to 64 steps

Haven't checked the code, but I suspect the current random gen. probably IS based on Perlin noise, or something conceptually very similar.

**Comment by jstefani** [comment]
I've been writing generators and the implementation seems good enough to me. My only question is if there is a reason that functions such as 

GeneratorPage::drawEuclideanGenerator

and

GeneratorPage::drawRandomGenerator

are inside ui/pages/GeneratorPage.cpp rather than inside their own classes and then called with a more generic drawSelf() function or similar in GeneratorPage.cpp

With the current implementation, GeneratorPage.cpp has to be modified for every new generator added. Not really a problem - just a comment.

**Comment by westlicht** [comment]
@jstefani yes, the current implementation is bad in terms of separation. The thing is that I never got around to actually finish the generators because I've never fully settled on a concept. I have a few ideas in which direction to take them, so the current state is more of a work in progress / hack. That's also why I have not documented the feature yet, as functionality might change completely once I settle on a design.

**Comment by jstefani** [comment]
I've considered using shift + track to open the context menu that the track perhaps?

**Comment by captainhooke88** [comment]
I really love the generator functionality and use it a lot as is.

Would be great to have a way to affect multiple parameters at once. For instance Gate & Note.

Anyway already powerful tool, wondering what it will finally look like once you decide on the final concept.

**Comment by vdgno** [comment]
I have ver 0.1.35. when entering random or euclidian generator from step screen, all sequencer buttons stop to response except function buttons. Definitely a bug here.

**Comment by gitmusik** [comment]
I agree that the separation thing slows one down a little but I absolutely LOVE the Euclidean/Generator functions. I use them exhaustively and find them superior in every way to anything else on any other sequencers to my knowledge.

**Comment by westlicht** [comment]
I should finally put in some work in generators. It's a great feature and I have some more ideas for other algorithms. But the UX could be better, I know that.

**Comment by modularev** [comment]
@westlicht was there a change which broke the function keys inside the `EUCLIDEAN ` and `RANDOM ` mode? `INIT` `REVERT` and `COMMIT` work. Also `EUCLIDEAN\STEPS` and `RANDOM\SEED` do, but none of the other options)
